[AMD] Enable masked load and pointer canonicalization pass

giuseros opened the pull request:
This PR is doing two things:
- We are using the new `llvm.masked{load/store}` intrinsics. This means that the backend will take responsibility to lower the stores/loads. 
- We are enabling the canonicalization pointer pass on the Triton IR. I extensively run testing and corrected a couple of minor issues still present in the implementation. 

The reason why I am enabling both at the same time is because I saw a minor regression with `llvm.masked{load,store}` which seems to go away when using the pointer canonicalization. Also, this combination seems to reduce the numbers of vgprs used (at least for GEMM kernels).

giuseros commented:
Upon further thought, I sank the emission of `masked.load` and `masked.store` in `llLoad` and `llStore` (respectively). So for now we will use masked intrinsics when we can use them

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -35,6 +37,20 @@ std::string mangleFunc(std::string name, Type type) {
+
+// Create a constant vector mask of length `vec` with the same `pred` value
+Value createVectorMaskFromPredicate(RewriterBase &rewriter, Location loc,
+                                    Value pred, int64_t vec) {
```
Nit: name this as `vecSize` to be clear?

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -157,6 +173,15 @@ Value llGetPid(Location loc, RewriterBase &rewriter, ModuleOp moduleOp,
              Value pred, Value falseVal, triton::CacheModifier cm) {
+
+  if (cm == triton::CacheModifier::CG || cm == triton::CacheModifier::NONE) {
+    int64_t vec = cast<VectorType>(elemTy).getNumElements();
```
Nit: name it as `vecSize`?

antiagainst reviewed (approved):
Cool thanks! Just two small nits.

antiagainst commented:
Could you fix the test and update the pull request message too?

giuseros commented:
Unfortunately the issue with the test revealed something more problematic of the pass. I might do a separate PR about this.

giuseros commented:
Hi @antiagainst , this is the PR that fixes the issue: https://github.com/triton-lang/triton/pull/4678

giuseros commented:
Hi @antiagainst , I rebased, fixed the nits and also addressed the fact that we are now using `llLoad` also for LDS access. This mean that I had to add support for non-vector types when I emit the masked.load. I ran few tests and everything seems ok (also the FA test). Let's see how the CI does.

giuseros commented:
Hi @antiagainst , could you start the tests again ?
